### PR TITLE
provisioner: Fix ssh key validation (SOC-11126)

### PR DESCRIPTION
**Pull request by skazi0**
Fixed ssh keys validation wrt (optional) options field before the
type/method in key line.
Additional improvement is proper handling of comments with spaces and
extended list of supported key types.

Parsing is based on rules described here: https://man.openbsd.org/sshd#AUTHORIZED_KEYS_FILE_FORMAT

Test input:
```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC5pApJQ6rC+KAMSeXSLhqJTTcyNKrl7EU41XYzJoj0/AmwNtAt4eVNwVgJ3q0Uvw8DUtSYrb6d6jQCzsm4qdpc7dk1u8lIIYOlv/QYveAnsfkQTslocXZBsgBNcEUNguVVutUWDux4q/O5y8VU7wIr+5KtnEuFNO6V1h0TrC6wP+TcHCC5NnESBj/RHUkQc22xuc19HR/1pJf9WrUz+1pSK8dD96ckPIs3H6E0U6uVYuqn7GT9KeETxPKYZhs2qtwS/BBxtZZ/BbXQtzF0yHkLPBnutdIwcG74TF7QAAsECTkAsMi0FdABJeLTQA/7i7tv/KNBiHmf0wH9eHuGdXHx skazi@linux-ofh6.suse

# Comments allowed at start of line
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H user@example.net
from="*.sales.example.net,!pc.sales.example.net" ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H john@example.net
command="dump /home",no-pty,no-port-forwarding ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H example.net
permitopen="192.0.2.1:80",permitopen="192.0.2.2:25" ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H
permitlisten="localhost:8080",permitopen="localhost:22000" ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H
tunnel="0",command="sh /etc/netstart tun0" ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H jane@example.net
restrict,command="uptime" ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H user1@example.net
restrict,pty,command="nethack" ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H user2@example.net
no-touch-required sk-ecdsa-sha2-nistp256@openssh.com AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H user3@example.net

environment="MYVAR=I have a quote\" in my middle" ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H user4@example.net

ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX3SgWvI+w3uKlffi6Xbl0s4KIHek3lg69qdGuDrSsevUzfkAug6T60XqJsTXAi2LvwdppXgWAORK/6eK2CKa/XGc4taC3nhO1p+iZ7ziG9uiAa9ei8JSrW68Lfmw8jjPP5F69v/batlrvlDoCoIL4sGSz6KljcvQJ/9wMMze48Umz+PcB5NIt11E95Onyov/esz2uhvnCXHH7Lsgq2HjPoXabNxEo2RNKFzH7DDf/BhPlJxpBNMxiUjTWQNet3Oq6hoZg7GPy772xClztm7TwrzoTT9OWK646+CTjer3mLXP10pjm6PcmdkObIceXD7hLgnMKdM6XAkFKdDt9wy6H comment with spaces user@example.net

invalid line which is not a key at all
invalid line four field
invalid line three
```

### CODE REVIEW

**Comment by skazi0** [comment]
before (part):
Zrzut ekranu z 2020-04-03 00-11-06

after:
Zrzut ekranu z 2020-04-03 00-08-37

(test input in PR description)

**Comment by jsuchome** [comment]
Hm, this looks like something that deserves a testsuite...

File: crowbar_framework/spec/lib/crowbar/ssh_key_parser_spec.rb [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      line = 'no-touch-required sk-ecdsa-sha2-nistp256@openssh.com %s user3@example.net' % DUMMY_KEY_CONTENTS
+      (options, type, key, comment) = DummyKeyParser.new.split_key_line(line)
+      expect(options).to eq("no-touch-required")
+      expect(type).to eq("sk-ecdsa-sha2-nistp256@openssh.com")
+      expect(key).to eq(DUMMY_KEY_CONTENTS)
+      expect(comment).to eq("user3@example.net")
+    end
+
+    it "with quotes and spaces in options" do
+      line = 'environment="MYVAR=I have a quote\" in my middle" ssh-rsa %s user4@example.net' % DUMMY_KEY_CONTENTS
```
[the remark this answers is not in the document]
Comment (skazi0) [reply crowbar_framework/spec/lib/crowbar/ssh_key_parser_spec.rb:112]: Dear Hound,
I really prefer '%s' over '%s'.
The `format` takes more space and you will complain about even longer lines.
This variable is called `line` for a reason. I'm not going to break it across multiple lines.
BR :oP

File: crowbar_framework/app/models/provisioner_service.rb [hunk 2]
```diff
@@ -57,11 +59,11 @@ def validate_proposal_after_save proposal
...
       key.strip!
 
-      next if key.empty?
+      next if key.empty? || key[0] == "#"
 
-      method, public_key, comment = key.split(" ")
-      unless ["ssh-rsa", "ssh-ed25519", "ssh-dss"].include? method
-        validation_error("SSH key \"#{key}\" should start with a supported \"ssh-*\" method.")
+      _, method, public_key, comment = split_key_line(key)
+      unless valid_key_types.include? method
```
Comment (jsuchome) [inline crowbar_framework/app/models/provisioner_service.rb:65]: So `valid_key_types` is accessible just like this, no module prefix? (Pardon my missing Ruby knowledge)

Comment (skazi0) [reply crowbar_framework/app/models/provisioner_service.rb:65]: We `include` the module into this class so yes, it is thrown directly into the class namespace. But (with my limited ruby knowledge) I had to spend quite some time to figure out how to do it "properly" :o)

**Pull request merged**